flink-runtime/src/main/java/org/apache/flink/runtime/heartbeat/HeartbeatManagerImpl.java

```diff
@@ -0,0 +1,328 @@
+	/**
+	 * Heartbeat monitor which manages the heartbeat state of the associated heartbeat target. The
+	 * monitor notifies the {@link HeartbeatListener} whenever it has not seen a heartbeat signal
+	 * in the specified heartbeat timeout interval. Each heartbeat signal resets this timer.
+	 *
+	 * @param <O> Type of the payload being sent to the associated heartbeat target
+	 */
+	static class HeartbeatMonitor<O> implements Runnable {
+
+		/** Resource ID of the monitored heartbeat target */
+		private final ResourceID resourceID;
+
+		/** Associated heartbeat target */
+		private final HeartbeatTarget<O> heartbeatTarget;
+
+		private final ScheduledExecutorService scheduledExecutorService;
+
+		/** Listener which is notified about heartbeat timeouts */
+		private final HeartbeatListener<?, ?> heartbeatListener;
+
+		/** Maximum heartbeat timeout interval */
+		private final long heartbeatTimeoutIntervalMs;
+
+		private volatile ScheduledFuture<?> futureTimeout;
+
+		private final AtomicReference<State> state = new AtomicReference<>(State.RUNNING);
+
+		HeartbeatMonitor(
+			ResourceID resourceID,
+			HeartbeatTarget<O> heartbeatTarget,
+			ScheduledExecutorService scheduledExecutorService,
+			HeartbeatListener<?, O> heartbeatListener,
+			long heartbeatTimeoutIntervalMs) {
+
+			this.resourceID = Preconditions.checkNotNull(resourceID);
+			this.heartbeatTarget = Preconditions.checkNotNull(heartbeatTarget);
+			this.scheduledExecutorService = Preconditions.checkNotNull(scheduledExecutorService);
+			this.heartbeatListener = Preconditions.checkNotNull(heartbeatListener);
+
+			Preconditions.checkArgument(heartbeatTimeoutIntervalMs >= 0L, "The heartbeat timeout interval has to be larger than 0.");
+			this.heartbeatTimeoutIntervalMs = heartbeatTimeoutIntervalMs;
+
+			resetHeartbeatTimeout(heartbeatTimeoutIntervalMs);
+		}
+
+		HeartbeatTarget<O> getHeartbeatTarget() {
+			return heartbeatTarget;
+		}
+
+		void reportHeartbeat() {
+			resetHeartbeatTimeout(heartbeatTimeoutIntervalMs);
+		}
+
+		void resetHeartbeatTimeout(long heartbeatTimeout) {
+			if (state.get() == State.RUNNING) {
+				cancelTimeout();
+
+				futureTimeout = scheduledExecutorService.schedule(this, heartbeatTimeout, TimeUnit.MILLISECONDS);
+
+				// Double check for concurrent accesses (e.g. a firing of the scheduled future)
+				if (state.get() != State.RUNNING) {
+					cancelTimeout();
+				}
+			}
+		}
+
+		void cancel() {
+			// we can only cancel if we are in state running
+			if (state.compareAndSet(State.RUNNING, State.CANCELED)) {
+				cancelTimeout();
+			}
+		}
+
+		private void cancelTimeout() {
+			if (futureTimeout != null) {
+				futureTimeout.cancel(true);
+			}
+		}
+
+		public boolean isCanceled() {
+			return state.get() == State.CANCELED;
+		}
+
+		@Override
+		public void run() {
+			// The heartbeat has timed out if we're in state running
+			if (state.compareAndSet(State.RUNNING, State.TIMEOUT)) {
+				heartbeatListener.notifyHeartbeatTimeout(resourceID);
```

Comment:
Took me a while to figure out this is called by the executor after a timeout. Makes sense.